### PR TITLE
(IAC-671) Implement New-PuppetDscModule

### PR DESCRIPTION
**Pull request by michaeltlombardi**
This PR adds the new public function, `New-PuppetDscModule`, which allows a module user to specify a single command and Puppetize a PowerShell module with DSC resources and is the main interface for the module.

It also updates the build script to call `New-PuppetDscModule` and remove all of the old build logic, leaving it intact as an interface for executing the code without duplicating code.

Finally, it privatizes several functions which are not useful in isolation, only during a module Puppetization:

- `Add-DscResourceModule`
- `Initialize-PuppetModule`
- `Update-PuppetModuleFixture`
- `Update-PuppetModuleMetadata`
- `Update-PuppetModuleReadme`

### CODE REVIEW

File: src/functions/New-PuppetDscModule.ps1 [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    If ([string]::IsNullOrEmpty($OutputDirectory))  {
+      $OutputDirectory  = Join-Path -Path (Get-Location) -ChildPath 'import'
+    } Else {}
+
+    $PuppetModuleRootFolderDirectory = Join-Path -Path $OutputDirectory                 -ChildPath $PuppetModuleName
+    $VendoredDscResourcesDirectory   = Join-Path -Path $OutputDirectory                 -ChildPath "$PuppetModuleName/lib/puppet_x/dsc_resources"
+    $PuppetModuleTypeDirectory       = Join-Path -Path $PuppetModuleRootFolderDirectory -ChildPath 'lib/puppet/type'
+    $PuppetModuleProviderDirectory   = Join-Path -Path $PuppetModuleRootFolderDirectory -ChildPath 'lib/puppet/provider'
+    $InitialPSModulePath          = $Env:PSModulePath
+    $InitialErrorActionPreference = $ErrorActionPreference
```
Comment (RandomNoun7) [inline src/functions/New-PuppetDscModule.ps1:61]: All of this alignment is a nice touch.

File: src/tests/functions/New-PuppetDscModule.Tests.ps1 [hunk 2]
```diff
@@ -0,0 +1,275 @@
+Describe "New-PuppetDscModule" {
+  InModuleScope puppet.dsc {
+    Context "Basic functionality" {
+      Mock Get-PuppetizedModuleName {$Name.ToLowerInvariant()}
+      Mock Initialize-PuppetModule {}
```
Comment (RandomNoun7) [inline src/tests/functions/New-PuppetDscModule.Tests.ps1:5]: I think Pester 5 might give us a way to mock at a higher level that will fall through child Contexts so you don't have to repeat this for every damn context block. That would be nice.

File: build.ps1 [hunk 3]
```diff
@@ -24,66 +24,13 @@
...
-
-  [string]$ProviderDirectoryPath  = Join-Path -Path $puppetProviderDir     -ChildPath $Resource.Name
-  [string]$ProviderFilePath       = Join-Path -Path $ProviderDirectoryPath -ChildPath $Resource.RubyFileName
-  if(-not(Test-Path $ProviderDirectoryPath)){
-    New-Item -Path $ProviderDirectoryPath -ItemType Directory -Force | Out-Null
-  }
-  [IO.File]::WriteAllLines($ProviderFilePath, $Resource.Provider)
-}
-$env:PSModulePath = $oldPsModulePath
+New-PuppetDscModule @PSBoundParameters
```
Comment (RandomNoun7) [inline build.ps1:36]: This is just really elegant.

**Comment by RandomNoun7** [comment]
@michaeltlombardi I'm ready to merge this if you are so you can hit the ground running Monday.

File: build.ps1 (hunk 3 above)
Comment (michaeltlombardi) [reply build.ps1:36]: Thanks! This is the last PR before the module functionality is exposed as a friendly UX, calling this function is _way_ better than trying to walk someone through the script, I think!

File: src/tests/functions/New-PuppetDscModule.Tests.ps1 (hunk 2 above)
Comment (michaeltlombardi) [reply src/tests/functions/New-PuppetDscModule.Tests.ps1:5]: This is one of those times I really wish pester was more like rspec, I would like it to inherit mock definitions but reset how many times it's been called, or track calls in this context separately or something.

**Pull request merged**